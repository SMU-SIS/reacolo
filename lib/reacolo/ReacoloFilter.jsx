import React, { PureComponent } from 'react';
import ComponentFilter from './ComponentFilter';
import contextPropType from './context-prop-type';
import valueGroupsFilter from './filtering/value-groups-filter';
import targetParser from './filtering/target-grammar.pegjs';

const parseTargets = (...args) => targetParser.parse(...args);

class ReacoloFilter extends PureComponent {
  static wouldRender(context, targets) {
    const { roles, activity, clientRole } = context;
    // Build the value / target map. Adapts the context values so that they are all an dictionary
    // of count.
    const valueTargetMap = [
      [(roles || {}), targets.roles],
      [(activity ? { [activity]: 1 } : {}), targets.activity],
      [(clientRole ? { [clientRole]: 1 } : {}), targets.clientRole]
    ];
    // If the decoded boolean is true,
    return valueTargetMap.every(([values, targetGroup]) => {
      if (targetGroup == null) return true;
      const coded = typeof targetGroup === 'string';
      // If the target group is encoded, decode it.
      const decodedTargetGroup = coded ? parseTargets(targetGroup) : targetGroup;
      // Check the filter on this target / value pair.
      return valueGroupsFilter(values, decodedTargetGroup);
    });
  }
  render() {
    const { roles, activity, clientRole, override } = this.props;
    let shouldRender = override;
    if (override == null) {
      shouldRender = this.constructor.wouldRender(
        this.props.context,
        { roles, activity, clientRole }
      );
    }
    return (
      <ComponentFilter rendered={shouldRender}>
        { this.props.children }
      </ComponentFilter>
    );
  }
}

/* eslint-disable react/no-unused-prop-types */
ReacoloFilter.propTypes = {
  context: contextPropType,
  children: React.PropTypes.node.isRequired,
  roles: React.PropTypes.string,
  activity: React.PropTypes.string,
  clientRole: React.PropTypes.string,
  override: React.PropTypes.bool
};
/* eslint-enable react/no-unused-prop-types */

ReacoloFilter.defaultProps = {
  context: undefined,
  roles: undefined,
  activity: undefined,
  clientRole: undefined,
  override: undefined
};

export default ReacoloFilter;
