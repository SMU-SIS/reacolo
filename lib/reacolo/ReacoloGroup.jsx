import React, { PureComponent } from 'react';

class ReacoloGroup extends PureComponent {
  selectFilter() {
    const props = this.props;
    const context = props.context;
    const children = React.Children.toArray(props.children);
    // Look for the first filters that pass.
    const selectedFilter = children
      .find((child) => {
        const childProps = Object.assign({}, child.props);
        // Remove the children from the considered properties (it is ignored, but still).
        delete childProps.children;
        return child.type.wouldRender(context, childProps);
      });
    // If no filter passed, look for a default filter.
    return selectedFilter || children.find(child => child.props.default);
  }
  render() {
    const selectedFilter = this.selectFilter();
    // Force the rendering of the selected filter and returns it.
    if (selectedFilter) {
      return React.cloneElement(selectedFilter, { override: true });
    }
    return null;
  }
}

export default ReacoloGroup;
