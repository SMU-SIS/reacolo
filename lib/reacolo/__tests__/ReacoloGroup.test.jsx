import React from 'react';
import renderer from 'react-test-renderer';
import ReacoloGroup from '../ReacoloGroup';
import ReacoloFilter from '../ReacoloFilter';

jest.mock('../ReacoloFilter');

beforeEach(() => {
  ReacoloFilter.mockClear();
  ReacoloFilter.prototype.render.mockImplementation(function render() {
    return <div>{this.props.children}</div>;
  });
});

describe('ReacoloGroup', () => {
  it('properly calls wouldRender on its child', () => {
    renderer.create(
      <ReacoloGroup context={{ roles: { test: 'test' } }}>
        <ReacoloFilter
          activity="activity1"
          clientRole="clientRole1"
          roles="roles1"
        >
          child filter 1
        </ReacoloFilter>
        <ReacoloFilter
          activity="activity2"
          clientRole="clientRole2"
          roles="roles2"
        >
          child filter 2
        </ReacoloFilter>
        <ReacoloFilter
          activity="activity3"
          clientRole="clientRole3"
          roles="roles3"
        >
          child filter 3
        </ReacoloFilter>
      </ReacoloGroup>
    );

    expect(ReacoloFilter.wouldRender.mock.calls).toEqual([
      [
        { roles: { test: 'test' } },
        { activity: 'activity1',
          clientRole: 'clientRole1',
          roles: 'roles1' }
      ], [
        { roles: { test: 'test' } },
        { activity: 'activity2',
          clientRole: 'clientRole2',
          roles: 'roles2' }
      ], [
        { roles: { test: 'test' } },
        { activity: 'activity3',
          clientRole: 'clientRole3',
          roles: 'roles3' }
      ]
    ]);
  });
  it('renders the first filters that passes', () => {
    ReacoloFilter.wouldRender.mockImplementation((_, { clientRole }) => clientRole === 'ok');
    const component = renderer.create(
      <ReacoloGroup context={{ roles: {} }}>
        <ReacoloFilter>Not to be rendered</ReacoloFilter>
        <ReacoloFilter clientRole="ok">To be rendered</ReacoloFilter>
        <ReacoloFilter clientRole="ok">Not to be rendered</ReacoloFilter>
      </ReacoloGroup>
    );
    expect(ReacoloFilter.mock.instances.length).toBe(1);
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('injects the override property to the child it renders', () => {
    ReacoloFilter.wouldRender.mockImplementation((_, { clientRole }) => clientRole === 'ok');
    renderer.create(
      <ReacoloGroup context={{ roles: {} }}>
        <ReacoloFilter testProp="bar">Not to be rendered</ReacoloFilter>
        <ReacoloFilter clientRole="ok" testProp="foo">To be rendered</ReacoloFilter>
      </ReacoloGroup>
    );
    expect(ReacoloFilter.mock.instances.length).toBe(1);
    expect(ReacoloFilter.mock.instances[0].props).toEqual({
      clientRole: 'ok',
      children: 'To be rendered',
      testProp: 'foo',
      override: true
    });
  });
  it('does not render if no filters would render and there is no default filter', () => {
    ReacoloFilter.wouldRender.mockReturnValue(false);
    const component = renderer.create(
      <ReacoloGroup context={{ roles: {} }}>
        <ReacoloFilter>Not to be rendered</ReacoloFilter>
        <ReacoloFilter>Not to be rendered</ReacoloFilter>
      </ReacoloGroup>
    );
    expect(ReacoloFilter.mock.instances.length).toBe(0);
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('renders the default filter if none would render', () => {
    ReacoloFilter.wouldRender.mockReturnValue(false);
    const component = renderer.create(
      <ReacoloGroup context={{ roles: {} }}>
        <ReacoloFilter>Not to be rendered</ReacoloFilter>
        <ReacoloFilter default>Default Filter Content</ReacoloFilter>
        <ReacoloFilter>Not to be rendered</ReacoloFilter>
      </ReacoloGroup>
    );
    expect(ReacoloFilter.mock.instances.length).toBe(1);
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('renders the filter that would render even if there is a default filter', () => {
    ReacoloFilter.wouldRender.mockImplementation((_, { clientRole }) => clientRole === 'ok');
    const component = renderer.create(
      <ReacoloGroup context={{ roles: {} }}>
        <ReacoloFilter>Not to be rendered</ReacoloFilter>
        <ReacoloFilter clientRole="ok">To be rendered</ReacoloFilter>
        <ReacoloFilter default>Default Filter Content</ReacoloFilter>
        <ReacoloFilter clientRole="ok">Not to be rendered</ReacoloFilter>
        <ReacoloFilter>Not to be rendered</ReacoloFilter>
      </ReacoloGroup>
    );
    expect(ReacoloFilter.mock.instances.length).toBe(1);
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('renders the filter that would render even it comes after the default filter', () => {
    ReacoloFilter.wouldRender.mockImplementation((_, { clientRole }) => clientRole === 'ok');
    const component = renderer.create(
      <ReacoloGroup context={{ roles: {} }}>
        <ReacoloFilter>Not to be rendered</ReacoloFilter>
        <ReacoloFilter default>Default Filter Content</ReacoloFilter>
        <ReacoloFilter clientRole="ok">To be rendered</ReacoloFilter>
        <ReacoloFilter clientRole="ok">Not to be rendered</ReacoloFilter>
      </ReacoloGroup>
    );
    expect(ReacoloFilter.mock.instances.length).toBe(1);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
