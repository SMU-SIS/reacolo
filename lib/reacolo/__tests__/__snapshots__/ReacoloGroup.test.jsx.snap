exports[`ReacoloGroup does not render if no filters would render and there is no default filter 1`] = `null`;

exports[`ReacoloGroup renders the default filter if none would render 1`] = `
<div>
  Default Filter Content
</div>
`;

exports[`ReacoloGroup renders the filter that would render even if there is a default filter 1`] = `
<div>
  To be rendered
</div>
`;

exports[`ReacoloGroup renders the filter that would render even it comes after the default filter 1`] = `
<div>
  To be rendered
</div>
`;

exports[`ReacoloGroup renders the first filters that passes 1`] = `
<div>
  To be rendered
</div>
`;
